addon/PageScraper.js

```diff
@@ -0,0 +1,77 @@
+/* globals Services, Task */
+"use strict";
+
+const {MetadataParser} = require("addon/MetadataParser");
+const {Cu} = require("chrome");
+const options = require("@loader/options");
+
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/Task.jsm");
+Cu.importGlobalProperties(["URL"]);
+
+const DEFAULT_OPTIONS = {
+  metadataTTL: 3 * 24 * 60 * 60 * 1000, // 3 days for the metadata to live
+  framescriptPath: new URL("data/page-scraper-content-script.js", options.prefixURI),
+  blackList: ["about:", "localhost", "resource://"]
+};
+
+function PageScraper(previewProvider, options = {}) {
+  this.options = Object.assign({}, DEFAULT_OPTIONS, options);
+  this._previewProvider = previewProvider;
+  this._metadataParser = new MetadataParser();
+}
+
+PageScraper.prototype = {
+  /**
+   * Check if the link is already in the metadata database. If not, parse the
+   *  HTML and insert it in the metadata database
+   */
+  _asyncParseAndSave: Task.async(function*(raw, url) {
+    let link = yield this._previewProvider.asyncDoesSingleLinkExist(url);
+    if (!link) {
+      let metadata = yield this._metadataParser.parseHTMLText(raw, url);
+      this._insertMetadata(metadata);
+    }
+  }),
+
+  /**
+   * Insert the metadata in the metadata database, along with it's source.
+   * Also add it to the metadata cache
+   */
+  _insertMetadata(metadata) {
+    if (Object.keys(metadata).length) {
+      this._previewProvider.processAndInsertMetadata([metadata], "Local");
```

Comment:
The check here is because if there is no metadata returned for that page, the service returns an empty object. Since we can't be computing cache keys and stuff if there's no ```url``` property, we don't want to even be doing the step of processing and inserting
I'll double check with @jaredkerim, maybe it throws and a try-catch is a better option